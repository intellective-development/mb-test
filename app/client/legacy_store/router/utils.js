// @flow

import _ from 'lodash';
import qs from '../../utils/qs';
import { encodeObject, decodeObject } from 'shared/utils/convert_utf8_b64';
import { filter_constants, filter_helpers } from 'store/business/filter';
import type { Filter } from 'store/business/filter';
import { product_list_constants } from 'store/business/product_list';

export function encodePLPParams(filter: $Shape<Filter>, sort?: product_list_constants.SortOptionId, base_url_filter: Object = {}){
  const param_filter = _.omit(filter, Object.keys(base_url_filter));

  // We protect the types of the more complex filter attrs options by encoding them with JSON
  // TODO: We shouldn't strictly need this, though there are non-string types in the filter.
  // We should consider a separate function to handle those on a case by case basis.
  const formatted_filter = _.mapValues(param_filter, (filter_value: Object, filter_attr: string) => {
    return JSON.stringify(filter_value);
  });

  // note that sort will not be included out if it is undefined, as qs ignoes undefined keys
  return qs.stringify({...formatted_filter, sort});
}

export function decodePLPParams(params: ?string){
  if (!params) return {};

  const { sort, ...rest_params } = qs.parse(params);
  const filter = _.pick(rest_params, filter_constants.FILTER_KEYS);

  // as we encode the Filter's selected attrs in the URL, we need to decode them here
  const parsed_filter = _.mapValues(filter, (filter_value: string, filter_attr: string) => {
    return JSON.parse(filter_value);
  });

  return { sort, filter: parsed_filter };
}

export const isExternallyNavigable = (destination: string) => { // TODO: test
  const is_brand_plp = /store\/brand\//.test(destination);
  const is_pdp = /store\/product\//.test(destination);

  return is_brand_plp || is_pdp;
};

export const navigateTo = function(destination: string, options: Object = {trigger: true}){
  const short_destination = window.Minibar.stripUrlBase(destination);
  window.Minibar.navigateOrReload(short_destination, options);
};