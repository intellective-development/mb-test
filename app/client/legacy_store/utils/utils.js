jQuery.fn.insertAt = function(index, element) {
  var lastIndex = this.children().size()
  if (index < 0) {
    index = Math.max(0, lastIndex + 1 + index)
  }
  this.append(element)
  if (index < lastIndex) {
    this.children().eq(index).before(this.children().last())
  }
  return this;
};


window.Validation = {
  errorPlacement: function(error, element) {
    var label = $('label[for=\''+$(element).attr('id')+'\']');
    label.hide().after(error);
  },
  highlight: function(element, errorClass, validClass) {
    $(element).addClass(errorClass).removeClass(validClass);
  },
  unhighlight: function(element, errorClass, validClass) {
    var el = $(element),
      label = $('label[for=\''+el.attr('id')+'\']');

    el.removeClass(errorClass).addClass(validClass);
    label.show();
  }
}


var backbone_sync = Backbone.sync, xhrPool = [];

Backbone.sync = function(method, model, options) {
  options = options || {};
  if (method === 'read') {
    if (options.abortPending === true) {
      for (var i = 0; i < xhrPool.length; i++) {
        if (xhrPool[i].readyState > 0 && xhrPool[i].readyState < 4) {
          xhrPool[i].abort();
          xhrPool.splice(i, 1);
        }
      }
    }

    for (var i = 0; i < xhrPool.length; i++) {
      if (xhrPool[i].readyState === 4) {
        xhrPool.splice(i, 1);
      }
    }

    var xhr = backbone_sync(method, model, options);
    xhrPool.push(xhr);
    return xhr;
  } else {
    return backbone_sync(method, model, options);
  }
};

export default {};
