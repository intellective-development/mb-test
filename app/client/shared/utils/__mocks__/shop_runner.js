const hasShopRunnerToken = jest.fn(() => false);

export { hasShopRunnerToken };
