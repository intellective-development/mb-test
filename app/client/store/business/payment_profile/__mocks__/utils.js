// @flow

export const tokenizeCard = jest.fn();
