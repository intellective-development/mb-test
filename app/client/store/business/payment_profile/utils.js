// @flow

import braintree_client from 'braintree-web/client';

type CardData = {
  number: string,
  expirationMonth: string,
  expirationYear: string,
  cvv: string,
  billingAddress: {
    postalCode: string
  }
}
type TokenizeOptions = {
  braintree_client_token: string
}

const BRAINTREE_ERROR_BODY = {message: 'Invalid Credit Card'};
export const tokenizeCard = (card_data: CardData, {braintree_client_token}: TokenizeOptions): Promise<string> => {
  const tokenize_card_prom = new Promise((resolve, reject) => {
    braintree_client.create(
      { authorization: braintree_client_token },
      (create_error, client_instance) => {
        if (create_error){
          if (global.Raven) global.Raven.captureMessage(create_error);
          reject({error: BRAINTREE_ERROR_BODY});
          return null;
        }
        return client_instance.request({
          endpoint: 'payment_methods/credit_cards',
          method: 'post',
          data: { creditCard: card_data, options: { validate: false } }
        }, (request_error, response) => {
          if (request_error){
            if (global.Raven) global.Raven.captureMessage(request_error);
            reject({error: BRAINTREE_ERROR_BODY});
          } else {
            const nonce = response.creditCards[0].nonce;
            resolve(nonce);
          }
        });
      }
    );
  });

  return tokenize_card_prom;
};
