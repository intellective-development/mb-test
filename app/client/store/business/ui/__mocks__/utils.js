// @flow

export const enterStore = jest.fn();
