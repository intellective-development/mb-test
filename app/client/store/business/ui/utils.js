// @flow

export const enterStore = (initial_destination: string = '/store') => {
  let formatted_destination = initial_destination;
  if (formatted_destination.substr(0, 1) !== '/'){ // append to root
    formatted_destination = `/${formatted_destination}`;
  }

  const prefix = '/store'; // formatted_destination will now always have a leading slash, don't need trailing here
  // if formatted_destination doesn't start with /store, prepend it. for promos
  if (formatted_destination.substr(0, prefix.length) !== prefix){
    formatted_destination = prefix + formatted_destination;
  }

  window.location.href = formatted_destination.toLowerCase();
};
