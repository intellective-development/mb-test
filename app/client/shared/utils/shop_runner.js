// @flow
import _ from 'lodash';
import * as mb_cookie from 'store/business/utils/mb_cookie';

// This returns a boolean based on SR login state. We are assuming if the `sr_token` cookie is present then the user is logged in.
const hasShopRunnerToken = (): boolean => {
  return mb_cookie.isPresent('sr_token');
};

// SR content is inserted by their partner integration kit. Since their divs are not present on DOM ready we need to tell the PIK to render content.
const refreshShopRunnerContent = () => {
  if (window.Data.shoprunner && _.isFunction(window.sr_updateMessages)) window.sr_updateMessages();
};

export { hasShopRunnerToken, refreshShopRunnerContent };
