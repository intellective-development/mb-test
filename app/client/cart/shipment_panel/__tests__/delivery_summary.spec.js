import * as React from 'react';
import dm_factory from '@minibar/store-business/src/delivery_method/__tests__/delivery_method.factory';
import supplier_factory from '@minibar/store-business/src/supplier/__tests__/supplier.factory';
import cart_item_factory from 'store/business/cart_item/__tests__/cart_item.factory';

import TestProvider from 'store/views/__tests__/utils/TestProvider';
import { hasShopRunnerToken } from 'client/shared/utils/shop_runner';
import DeliverySummary from '../delivery_summary';

jest.mock('client/shared/utils/shop_runner');

describe('DeliverySummary', () => {
  it('renders', () => {
    const delivery_methods = [dm_factory.build('on_demand'), dm_factory.build('shipped')];
    const shipment = {
      supplier: supplier_factory.build({delivery_methods}),
      delivery_method: delivery_methods[0],
      items: [cart_item_factory.build('five_dollar')]
    };

    expect(render(
      <TestProvider>
        <DeliverySummary shipment={shipment} delivery_method={delivery_methods[0]} />
      </TestProvider>
    )).toMatchSnapshot();
  });

  it('renders ShopRunner divs', () => {
    const delivery_methods = [dm_factory.build('on_demand'), dm_factory.build('shipped')];
    const shipment = {
      supplier: supplier_factory.build({delivery_methods}),
      delivery_method: delivery_methods[0],
      items: [cart_item_factory.build('five_dollar')]
    };

    hasShopRunnerToken.mockReturnValue(true);

    expect(render(
      <TestProvider>
        <DeliverySummary shipment={shipment} delivery_method={delivery_methods[0]} />
      </TestProvider>
    )).toMatchSnapshot();
  });
});
