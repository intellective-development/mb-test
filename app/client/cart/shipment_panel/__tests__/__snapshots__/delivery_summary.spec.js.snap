// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliverySummary renders 1`] = `
<div
  class="el-touchable el-touchable--disabled sp__dm__container"
  role="button"
  tabindex="-1"
>
  <div
    class="el-radio__circle-outer sp__dm__radio"
  >
    <div
      class="el-radio__circle-inner"
    />
  </div>
  <div
    class="sp__dm__name-expectation-container"
  >
    <span
      class="elMBText elMBText_ResetSpacing sp__dm__type"
    >
      Local Delivery
    </span>
    <span
      class="elMBText elMBText_ResetSpacing sp__dm__next_delivery"
    >
      <span>
        <span
          class="sp__dm__closed"
        >
          Closed
        </span>
         - 
      </span>
      Next available: 1 hour or less
    </span>
  </div>
  <span
    class="elMBText elMBText_ResetSpacing sp__dm__cost sp__dm__cost--minimum"
  >
    $25 Minimum
  </span>
</div>
`;

exports[`DeliverySummary renders ShopRunner divs 1`] = `
<div
  class="el-touchable el-touchable--disabled sp__dm__container"
  role="button"
  tabindex="-1"
>
  <div
    class="el-radio__circle-outer sp__dm__radio"
  >
    <div
      class="el-radio__circle-inner"
    />
  </div>
  <div
    class="sp__dm__name-expectation-container"
  >
    <span
      class="elMBText elMBText_ResetSpacing sp__dm__type"
    >
      Local Delivery
    </span>
    <span
      class="elMBText elMBText_ResetSpacing sp__dm__next_delivery"
    >
      <span>
        <span
          class="sp__dm__closed"
        >
          Closed
        </span>
         - 
      </span>
      Next available: 1 hour or less
    </span>
  </div>
  <div
    style="display:flex;flex-grow:1;text-align:right;min-width:200px"
  >
    <div
      style="width:100%"
    >
      <div
        class="_SRD"
        style="display:inline-block;vertical-align:middle;margin-right:0.5em"
      >
        <div
          class="srd_iconline"
        >
          <div
            class="srd_logo"
          />
        </div>
      </div>
      Free delivery
    </div>
  </div>
</div>
`;
